lightning/src/ln/peer_handler.rs

```diff
@@ -673,6 +708,25 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref> PeerManager<D
 		}
 	}
 
+	/// Find the peer corresponding to the given node id, and if found appends
+	/// a custom message to the outbound/write buffer.
+	pub fn enqueue_custom_message<M: Encode + Writeable + Debug>(&self, node_id: &PublicKey, message: &M) {
```

Comment:
This seems really hard to use correctly - right now if you receive a message via `handle_custom_message` and are in the midst of processing it, you cannot call `enqueue_custom_message` or you'll deadlock. This seems like precisely *the* way users would implement this API, though. I think the only other way you can reasonably go about this would be to add the ability to get the message id from the `CustomMessage` associated type (which is already going to be an enum.

Reply:
Oh that's a really good point.

Though I must say I don't understand what you would like to do instead. I tried several things previously including reusing the associated type for sending, but IIRC the main issue is the `const` requirement on the `Encode` trait, which makes it impossible to implement for an enum (or anything that has more than one type id).
I had also tried some complicated things like passing callbacks around but it always ended up requirement some boxing which didn't feel really nice.

For now I added a (mutexed) vector of node id + encoded messages on the peer handler, that can be used by the custom handler through the `enqueue_custom_message` method. It's a tiny bit a shame to have this while it might be mostly unneeded, but it felt like the simplest solution. Let me know what you think.

Reply:
>  tried several things previously including reusing the associated type for sending, but IIRC the main issue is the const requirement on the Encode trait, which makes it impossible to implement for an enum (or anything that has more than one type id).

Yep, I think the only practical solution here is to create a new trait for public consumption that doesn't use the const. I suppose we could also drop the const in Encode and swap it for just a method, but that's a lot more diff and I dunno if its worth it.

Reply:
I added the `TypedMessage` trait and made a blanket implementation for `T: Encode` so that the `write` function can be used by both. Because of that `Encode` needs to be public but apart from that I think that's a good way to reduce the diff.

I also updated the `CustomMessageHandler` to have a `get_and_clear_pending_msg` method so that it's more consistent with other handlers.